src/module_4/notebook.md

```diff
@@ -0,0 +1,2102 @@
+    Val until:  2021-02-22 00:00:00
+    Test until:  2021-03-03 00:00:00
+
+
+
+```python
+train_df = df_selected[df_selected.order_date <= train_val_cutoff]
+val_df = df_selected[(df_selected.order_date > train_val_cutoff) & (df_selected.order_date <= val_test_cutoff)]
+test_df = df_selected[df_selected.order_date > val_test_cutoff]
+```
+
+### Custom functions
+
+
+```python
+def plot_roc_and_pr_curves(model_name: str, y_test: pd.Series, y_pred: pd.Series, threshold_target: float=None, figure:Tuple[matplotlib.figure.Figure, np.array]=None):
+    """
+    Plots ROC and Precision-Recall curves side-by-side with AUC as legend.
+
+    Args:
+        model_name (str): The name of the classification model.
+        y_test (array-like): True labels.
+        y_preds (array-like): Predicted probabilities or scores.
+        precision_target (float, optional): Desired precision level for highlighting.
+        fig (matplotlib.figure.Figure, optional):  Figure object to plot on. 
+                                                  Creates a new figure if not provided.
+    """
+
+    if figure is None:
+        fig, ax = plt.subplots(1, 2, figsize=(12, 5))
+    else:
+        fig, ax = figure
+
+    # ROC Curve
+    fpr, tpr, roc_th = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+    ax[0].plot(fpr, tpr, label=f'{model_name} (AUC = {roc_auc:.2f})')
+    ax[0].set_xlabel('False Positive Rate')
+    ax[0].set_ylabel('True Positive Rate')
+    ax[0].set_title('ROC Curve')
+    ax[0].legend()
+
+    # Precision-Recall Curve
+    precision, recall, pr_th = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall, precision)
+    ax[1].plot(recall, precision, label=f'{model_name} (AUC = {pr_auc:.2f})')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve')
+    ax[1].legend()
+
+    # Highlight precision target (if provided)
+    if threshold_target is not None:
+        idx_0 = np.argmin(np.abs(roc_th - threshold_target))
+        ax[0].scatter(fpr[idx_0], tpr[idx_0])
+        idx_1 = np.argmin(np.abs(pr_th - threshold_target))
+        ax[1].scatter(recall[idx_1], precision[idx_1]) 
+```
+
+
+```python
+def auc_score(y_test: pd.Series, y_pred: pd.Series) -> Tuple[float, float]:
+    """
+    Returns the ROC and PR curves AUC score for a given set of true labels and predicted probabilities.
+
+    Args:
+        y_test (array-like): True labels.
+        y_preds (array-like): Predicted probabilities or scores.
+
+    Returns:
+        Tuple (float, float): ROC AUC score, PR curve AUC score.
+    """
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    return roc_auc, pr_auc
+```
+
+### Baseline
+
+We first try our baseline without any ML involved, just using a single column, `global_popularity` to predict the outcome.
```

Comment:
good idea to always use a baseline!

Reply:
however... is "total popularity" calculated over the WHOLE data set? does this mean that you are leaking data from the future in the past? I would double check that because otherwise this baseline is not totally correct